src/node.cc

```diff
@@ -3746,62 +3746,39 @@ Environment* CreateEnvironment(Isolate* isolate,
 }
 
 
-int Start(int argc, char** argv) {
-  PlatformInit();
-
-  const char* replaceInvalid = secure_getenv("NODE_INVALID_UTF8");
-
-  if (replaceInvalid == nullptr)
-    WRITE_UTF8_FLAGS |= String::REPLACE_INVALID_UTF8;
-
-  CHECK_GT(argc, 0);
-
-  // Hack around with the argv pointer. Used for process.title = "blah".
-  argv = uv_setup_args(argc, argv);
-
-  // This needs to run *before* V8::Initialize().  The const_cast is not
-  // optional, in case you're wondering.
-  int exec_argc;
-  const char** exec_argv;
-  Init(&argc, const_cast<const char**>(argv), &exec_argc, &exec_argv);
-
-#if HAVE_OPENSSL
-  // V8 on Windows doesn't have a good source of entropy. Seed it from
-  // OpenSSL's pool.
-  V8::SetEntropySource(crypto::EntropySource);
-#endif
-
-  V8::InitializePlatform(new Platform(4));
-
-  int code;
-  V8::Initialize();
-
-  // Fetch a reference to the main isolate, so we have a reference to it
+// Entry point for new node instances, also called directly for the main
+// node instance.
+void StartNodeInstance(void *arg) {
+  NodeInstanceData* instance_data = static_cast<NodeInstanceData*>(arg);
+  Isolate* isolate = Isolate::New();
+    // Fetch a reference to the main isolate, so we have a reference to it
   // even when we need it to access it from another (debugger) thread.
-  node_isolate = Isolate::New();
+  if (instance_data->is_main())
+    node_isolate = isolate;
   {
-    Locker locker(node_isolate);
-    Isolate::Scope isolate_scope(node_isolate);
-    HandleScope handle_scope(node_isolate);
-    Local<Context> context = Context::New(node_isolate);
+    Locker locker(isolate);
+    Isolate::Scope isolate_scope(isolate);
+    HandleScope handle_scope(isolate);
+    Local<Context> context = Context::New(isolate);
     Environment* env = CreateEnvironment(
-        node_isolate,
-        uv_default_loop(),
+        isolate,
+        instance_data->loop(),
         context,
-        argc,
-        argv,
-        exec_argc,
-        exec_argv);
+        instance_data->argc(),
+        instance_data->argv(),
+        instance_data->exec_argc(),
+        instance_data->exec_argv());
     Context::Scope context_scope(context);
-    env->set_using_abort_on_uncaught_exc(abort_on_uncaught_exception);
+    if (instance_data->is_main())
+      env->set_using_abort_on_uncaught_exc(abort_on_uncaught_exception);
     // Start debug agent when argv has --debug
-    if (use_debug_agent)
+    if (instance_data->is_main() && use_debug_agent)
```

Comment:
It's arguably cleaner to add a use_debug_agent() getter and configure it in Start().